keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,645 @@
+  the future, or to allow 3rd party validators to have a way to inline their
+  validation rules in CRDs.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Type Checking
+
+CEL type checking requires "declarations" be registered for any types that are to
+be type checked.  In our case, the type information of interest is the CRD's
+structural schemas. So we need to translate structural schemas to declarations.
+
+"declarations" are registered via the go-genproto "checked" package
+(https://github.com/googleapis/go-genproto/blob/master/googleapis/api/expr/v1alpha1/checked.pb.go).
+
+(https://github.com/googleapis/googleapis/blob/master/google/api/expr/v1alpha1/checked.proto).
+
+There are a couple alternative ways to do this. Ideally, we could be able to both dereference into
+objects and construct objects in a typesafe way.  In order to construct objects in a typesafe way we
+need to be able to represent the structural schema types in CEL, e.g. "v1beta1.Foo{fieldname:
+value}", this is complicated by the way CEL relies on protobuf types.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: CustomResourceValidationExpressions
+  - Components depending on the feature gate: kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+No, default behavior is the same.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, disabling the feature will result in validation expressions being ignored.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Validation expressions will be enforced again.
+
+###### Are there any tests for feature enablement/disablement?
+
+These will be introduced in the Alpha implementation.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+`x-kubernetes-validator` it not currently allowed in the OpenAPI schemas defined in Custom Resource
+Definitions. This creates a rollout issue: Any CRDs that are defined using this new field will
+be invalid according to versions of Kubernetes that pre-date the introduction of the field.
+
+
+Mitigation: Once we introduce the field, also backport the code that allows it to be included (but
+ignored) in CRDs to all supported Kubernetes versions. Before this feature goes to Beta we will
```

Comment:
please be sure to explicitly include this in the beta criteria.  Ending up with un-updateable resources on a downgrade would be rough.  For alpha, I think it's ok

Reply:
👍 Added Graduation Criteria section that includes this.